src/cacao.rs

```diff
@@ -0,0 +1,341 @@
+};
+use capgrok::{extract_capabilities, verify_statement, Error as CapGrokError, RESOURCE_PREFIX};
+use libipld::{cbor::DagCborCodec, codec::Decode};
+use serde::{Deserialize, Serialize};
+use std::convert::TryInto;
+use thiserror::Error;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(untagged)]
+pub enum HolderBindingDelegation {
+    // TODO
+    // IRIRef(crate::rdf::IRIRef),
+    Base64Block(String),
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Invalid Base64 Block")]
+    InvalidBase64Block,
+    #[error("JWT VCs not supported")]
+    UnsupportedJwtVc,
+    #[error(transparent)]
+    Base64(#[from] base64::DecodeError),
+    #[error(transparent)]
+    SiweVerification(#[from] SiweVerificationError),
+    #[error(transparent)]
+    Ipld(#[from] libipld::error::Error),
+    #[error(transparent)]
+    CapGrok(#[from] CapGrokError),
+}
+
+impl From<SIWEPayloadConversionError> for Error {
+    fn from(e: SIWEPayloadConversionError) -> Error {
+        SiweVerificationError::from(e).into()
+    }
+}
+
+impl HolderBindingDelegation {
+    /// Given a presentation id, presented credentials and a holder, return authorised holders
+    pub async fn validate(
+        &self,
+        id: Option<&URI>,
+        credentials: Option<&OneOrMany<CredentialOrJWT>>,
+        holder: Option<&URI>,
+    ) -> Result<Option<String>, Error> {
+        let HolderBindingDelegation::Base64Block(base64_cacao) = self;
+        let cbor_cacao = if let Some(b) = base64_cacao.strip_prefix("data:;base64,") {
+            base64::decode(b)?
+        } else {
+            return Err(Error::InvalidBase64Block);
+        };
+        let cacao: SiweCacao = CACAO::decode(DagCborCodec, &mut std::io::Cursor::new(cbor_cacao))?;
+        cacao.verify().await?;
+        let payload = cacao.payload();
+        let delegator = &payload.iss;
+        let delegee = &payload.aud;
+
+        if let Some(h) = holder {
+            if *delegee != h.to_string() {
+                return Ok(None);
+            }
+        } else {
+            return Ok(None);
+        }
+
+        let siwe: siwe::Message = payload.clone().try_into()?;
+        if !verify_statement(&siwe)? {
+            return Ok(None);
+        }
+        let capability = extract_capabilities(&siwe)?.remove(&"credentials".parse()?);
+
+        match (credentials, capability) {
+            (Some(credentials), Some(cap)) if !credentials.is_empty() => {
+                for credential in credentials.into_iter() {
+                    let (id, subjects) = match credential {
+                        CredentialOrJWT::Credential(c) => (c.id.as_ref(), &c.credential_subject),
+                        _ => return Err(Error::UnsupportedJwtVc),
+                    };
+                    if let Some(i) = id {
+                        // if the credential isnt presentable according to the siwe cap
+                        if !cap.can(i.as_str(), "present") {
+                            return Ok(None);
```

Comment:
I wonder about those returns, should the VC be either skipped, or return an error instead?

Reply:
In my understanding, it returns none to signify that there are no holders authorised to present the credential(s), while errors signify that the validation couldn't be completed

Reply:
Oh my bad I think I believed that the holder only needed to be authorised by one credential, or something along these lines. But it makes much more sense to enforce on all VCs.